src/solvers/flattening/arrays.cpp

```diff
@@ -185,9 +189,10 @@ void arrayst::collect_arrays(const exprt &a)
   }
   else if(a.id()==ID_member)
   {
-    if(to_member_expr(a).struct_op().id()!=ID_symbol)
-      throw
-        "unexpected array expression: member with `"+a.op0().id_string()+"'";
+    DATA_INVARIANT(
+      to_member_expr(a).struct_op().id()==ID_symbol,
+      ("unexpected array expression: member with `"+
+       a.op0().id_string()+"'").c_str());
```

Comment:
I think that looks good. It also gives the option to use an immediately-evaluated lambda as the `REASON` parameter, in the case that `ostringstream` is required to build the message string.

As you said, depending on the implementation of `invariant_failed` it might be better to do
```c++
std::string reason = (REASON); \
invariant_failed(__FILE__, __FUNCTION__, __LINE__, std::move(reason)); \
```